backends/backend_openssl3.c

```diff
@@ -2472,3 +2472,627 @@ static void openssl_ecdsa_backend(void)
+	size_t n_len = 0, e_len = 0;
+	(void)parsed_flags;
+	BN_CTX *bn_ctx = NULL;
+	bn_ctx = BN_CTX_new();
+	logger_binary(LOGGER_DEBUG, data->n.buf, data->n.len, "N");
+	logger_binary(LOGGER_DEBUG, data->d.buf, data->d.len, "D");
+
+	ret = rsa_create_pkey_crt(&pkey, data->n.buf, data->n.len, data->e.buf, data->e.len, data->u.rsa_crt.p.buf, data->u.rsa_crt.p.len,
+							data->u.rsa_crt.q.buf, data->u.rsa_crt.q.len, data->u.rsa_crt.dmp1.buf, data->u.rsa_crt.dmp1.len, 
+							data->u.rsa_crt.dmq1.buf, data->u.rsa_crt.dmq1.len, data->u.rsa_crt.iqmp.buf, data->u.rsa_crt.iqmp.len, bn_ctx);
+	CKNULL(pkey, -ENOMEM);
+	if(!ret) {
+		printf("pkey generation failed\n");
+		goto out;
+	}
+
+	CKNULL_LOG(pkey, -EFAULT, "Cannot allocate PKEY\n");
+
+	pkey_get_bn_bytes(pkey, OSSL_PKEY_PARAM_RSA_N,  &n, &n_len);
+	pkey_get_bn_bytes(pkey, OSSL_PKEY_PARAM_RSA_E, &e, &e_len);
+	logger_binary(LOGGER_DEBUG, n, n_len, "N");
+	logger_binary(LOGGER_DEBUG, e, e_len, "E");
+
+	ctx = EVP_PKEY_CTX_new_from_pkey(NULL, pkey, "");
+
+	CKNULL_LOG(ctx, -EFAULT, "Cannot allocate PKEY context\n");
+	CKINT_O_LOG(EVP_PKEY_encrypt_init(ctx), "PKEY decrypt init failed\n");
+
+	CKINT_O_LOG(EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_NO_PADDING),
+		    "Disabling padding failed\n");
+
+	/* Determine buffer length */
+	CKINT_O_LOG(EVP_PKEY_encrypt(ctx, NULL, &outlen, data->msg.buf,
+					data->msg.len),
+			"Getting Ciphertext length failed\n");
+
+	CKINT(alloc_buf(outlen, &data->signature));
+
+	ret = EVP_PKEY_encrypt(ctx, data->signature.buf, &outlen, data->msg.buf,
+				data->msg.len);
+
+	if (ret == 1) {
+		logger(LOGGER_DEBUG, "signature successful\n");
+	} else {
+		logger(LOGGER_DEBUG, "signature failed %s\n",
+			ERR_error_string(ERR_get_error(), NULL));
+	}
+	ret = 0;
+
+out:
+	if (pkey)
+		EVP_PKEY_free(pkey);
+	if (ctx)
+		EVP_PKEY_CTX_free(ctx);
+	return ret;
+}
+
+static int
+openssl_rsa_decryption_primitive(struct rsa_decryption_primitive_data *data,
+				flags_t parsed_flags)
+{
+	EVP_PKEY_CTX *ctx = NULL;
+	EVP_PKEY *key = NULL;
+	size_t outlen;
+	int ret;
+
+	(void)parsed_flags;
+
+	key = data->privkey;
+	CKNULL_LOG(key, -EFAULT, "Cannot allocate PKEY\n");
+	ctx = EVP_PKEY_CTX_new_from_pkey(NULL, key, "");
+	CKNULL_LOG(ctx, -EFAULT, "Cannot allocate PKEY context\n");
+
+	CKINT_O_LOG(EVP_PKEY_decrypt_init(ctx), "PKEY decrypt init failed\n");
+
+	CKINT_O_LOG(EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_NO_PADDING),
+			"Disabling padding failed\n")
+
+	/* Determine buffer length */
+	CKINT_O_LOG(EVP_PKEY_decrypt(ctx, NULL, &outlen, data->msg.buf,
+					data->msg.len),
+			"Getting plaintext length failed\n");
+
+	CKINT(alloc_buf(outlen, &data->s));
+
+	ret = EVP_PKEY_decrypt(ctx, data->s.buf, &outlen, data->msg.buf,
+				data->msg.len);
+	if (ret == 1) {
+		logger(LOGGER_DEBUG, "Decryption successful\n");
+		data->dec_result = 1;
+	} else {
+		logger(LOGGER_DEBUG, "Decryption failed %s\n",
+			ERR_error_string(ERR_get_error(), NULL));
+		data->dec_result = 0;
+	}
+
+	ret = 0;
```

Comment:
Why is ret set to 0 explicitly instead of assigning from return of a function call?